### ISSUE TITLE
Unfurl links to images in build artifacts

### ISSUE THREAD

**Issue by carlin-q-scott** [kiwicom/gitlab-unfurly#20]
I pasted a link like https://gitlab.com/my-project/-/jobs/1/artifacts/file/Reports/image.png and I expected it to be unfurled as an image. It was not unfurled.

Is this easy to add support for? I know most of the unfurling is just text-based.

**Comment by aexvir** [comment]
👋 Hi @carlin-q-scott 
Unfortunately I'd say this isn't possible without having to move the resources to somewhere else.

GitLab's [artifacts visibility is tied to the project's visibility](https://docs.gitlab.com/ee/ci/pipelines/settings.html#visibility-of-pipelines), so, for artifacts to be accessible from outside of GitLab the project would need to be public. And if that's the case, Slack's unfurling should already take care of downloading the image and showing it as metadata in Slack.

If the project isn't public, I'd say the metadata won't be able to be show anything, as there is no way for Slack's client to authenticate to GitLab to get the file. The only possible solution I could see here would be to move the files somewhere else, and then this bot could resolve the image to that one, but we aren't really interested into this at the moment.

One thing that could help here would be if [this issue](https://gitlab.com/gitlab-org/gitlab/-/issues/17544) is addressed from GitLab's side

**Comment by carlin-q-scott** [comment]
I appreciate your thorough feedback, and finding the related issue.

I can access the artifact file via the api using the access_token query param without being logged into gitlab. Perhaps the url can be converted to the api call and [attached to the message](https://api.slack.com/messaging/composing/layouts#attachments).

For example, this is a conversion from an artifact link to an api call. The :variable's would need to be filled in.
`https://gitlab.example.com/:group/:folder/:project/-/jobs/:jobid/artifacts/raw/:artifactpath`
`https://gitlab.example.com/api/v4/projects/:group%2F:folder%2F:project/jobs/:jobid/artifacts/:artifactpath?access_token=:access_token`
reference: https://docs.gitlab.com/ee/api/jobs.html#download-a-single-artifact-file-by-job-id

I tested the converted url with one of my private job artifacts, and the browser opens it for downloading, not displaying. That means it has the response header of `Content-Disposition:attachment`. I am unsure if a gitlab image attachment can work with such a Content-Disposition.

**Comment by aexvir** [comment]
Well, yeah, adding the access_token to the URL could work, but that would expose the token to anyone 😕 

Deploy tokens don't seem to be able to access artifacts based on the documentation, so the token used there would have API access and be able to do 
almost anything, as GitLab's token scoping is quite limited in this regard...

I really don't think this is a good solution

**Comment by carlin-q-scott** [comment]
Is the attachment URL exposed to the user? I don't think it is but I could be wrong.

**Comment by aexvir** [comment]
Yes it is...
<img width="1128" alt="Screenshot 2020-04-08 at 09 12 06" src="https://user-images.githubusercontent.com/8055505/78755205-12b3d300-7979-11ea-9a8d-5c5409d158dc.png">

**Comment by aexvir** [comment]
Yes it is, both from the API and the UI

```json
```

**Comment by aexvir** [comment]
And sorry, as we don't need it and, as I explained, there are security risks, we aren't considering this at all, I'll close this issue.

I'd be happy to review a PR if you find a way to do it without exposing any sensitive token. I suggested moving the files somewhere else, like a public s3 bucket. In that case it could be added as an optional functionality to this bot.

Up to you how you want to handle your public artifacts
